[Jed] Step 3-2: 리팩토링 & 속성 변경 동작 구현

junu0516 opened the pull request:
<img src="https://user-images.githubusercontent.com/68586291/156718581-c14d8940-7075-4cb2-90e6-bce36b785c0d.gif" style="width:30%; height:20%;"/>

## 작업 내용
- [x] Plane 구조체 생성 및 유닛테스트
    - [x] 생성한 사각형 모델 Plane에 추가
    - [x] 사각형 전체 개수 알려주는 computed property 선언  
    - [x] 인덱스 및 터치 좌표 넘겨서 사각형 모델을 리턴하는 Subscript 정의 
- [x] 사각형 터치할 때 해당 위치에 사각형 있으면 테두리 효과로 선택 표시
- [x] 사각형 선택했을 때 선택된 사각형의 배경색 및 투명도 화면 우측에 표시
- [x] 선택한 사각형의 투명도를 우측 슬라이더로 조정
- [x] 선택한 사각형의 배경색을 버튼 터치를 통해 랜덤으로 변경   
    
## 학습 키워드
- Protocol
- Subscript
- UIView, UIColor, UIGestureRecognizer
- Escaping Closure
    
## 고민과 해결
- 기존 1단계 피드백을 반영해서 Rectangle.Color 내부의 rgb의 범위를 0~255로 제한하는 로직을 추가했습니다.
    - 모든 Double 타입을 생성자 파라미터로 받을 수 있도록 하되, 0 미만의 값은 모두 0으로, 255를 초과하는 값은 모두 255로 강제로 수정하도록 하였습니다.
    
 - CanvasView, StylerView의 수정 혹은 Plane 내부의 Rectangle 객체의 값 수정 등은 모두 ViewController을 매개하여 일어나도록 했습니다.
     - CanvasView, StylerView 각각의 View가 ViewController을 내부에서 참조하되, UIViewController타입이 아닌 각각 ViewMutable, ModelMutable 프로토콜 타입으로 참조하도록 했습니다.
     - 각각의 View가 ViewController의 모든 부분을 알고 있기 보단, 필요한 부분만 알고 있는 것이 낫기에 별도의 프로토콜 타입으로 ViewController을 참조하는 것이 나을 것이라 생각했습니다.
     - StylerView가 슬라이더 혹은 배경색 버튼을 통해 디자인을 변경하면 ViewMutable 프로토콜의 메소드를 호출하여 ViewController가 CanvasView의 사각형 뷰 디자인을 바꾸는 메소드를 호출하도록 했습니다.
     - 반대로, CanvasView에서 배경색 혹은 투명도 등의 디자인이 변경되면 ModelMutable 프로토콜의 메소드를 호출하여 ViewController가 Plane의 Rectangle 모델 값을 변경하는 메소드를 호출하도록 했습니다.
     
- Plane, CanvasView가 모두 현재 선택된 사각형 관련 정보를 가진 프로퍼티를 가지도록 했습니다.
    - CanvasView의 경우 현재 선택된 SubView를 별도의 selectedRectangleView 프로퍼티가 참조하도록 했습니다.
    - Plane의 경우에는 현재 선택된 사각형에 대응되는 Rectangle 모델의 id값을 selectedRectangleId 프로퍼티가 참조하도록 했습니다.

junu0516 committed:
Refactor: Rectangle 클래스 id 프로퍼티 타입 수정

- Rectangle.Id 내부 타입 선언
- id 프로퍼티 타입을 String이 아닌 별도의 Rectangle.Id 타입으로 변경
- RectangleFactory에서 랜덤 문자열 생성하는 기능을 Rectangle.Id의 생성자로 이동

junu0516 committed:
Refactor: RectangleFactory 수정

- createRandomStringId() 메소드 내부의 랜덤 문자열 도출하는 로직 제거하고 리턴타입을 Rectangle.Id로 변경
- 팩토리 클래스의 asciiValues 프로퍼티 제거

junu0516 committed:
Refactor: Rectangle.Color 내부 프로퍼티 Double 값 변경

- 프로퍼티의 기본값들을 모두 0으로 설정
- 생성자에서 r,g,b 값을 받아서 모두 0~255사이의 수인지 검증한 값을 초기화
- 0 미만 값은 강제로 0으로 설정하고, 255 초과하는 값도 강제로 255로 설정

junu0516 committed:
Refactor: Rectangle.Alpha 열거형에서 구조체로 변경

- opacity 정수값을 생성자 파라미터로 받아서, 1~10 사이의 범위로 조정 후 프로퍼티값을 초기화하는 구조체로 변경
- 프로퍼티 opacity의 기본값은 완전 불투명 상태에 해당하는 10으로 우선 설정

junu0516 committed:
Refactor: RectangleFactory 클래스 내부 메소드 수정

- createRandomAlpha()의 리턴을 열거형이 아닌 구조체 인스턴스를 리턴하는 식으로 수정

junu0516 committed:
Add: Plane.swift 추가

junu0516 committed:
Revert "Refactor: Rectangle.Alpha 열거형에서 구조체로 변경"

This reverts commit 58c8711c37ecbf378bac41451d9942d741c1dd33.

junu0516 committed:
Revert "Refactor: RectangleFactory 클래스 내부 메소드 수정"

This reverts commit d2d3aea3a7321df10d3c4464d709e4f3203ca03a.

junu0516 committed:
Refactor: Rectangle.Alpha 열거형 내 opacity 인스턴스 추가

- 투명도값을 리턴하는 opacity 인스턴스를 추가

junu0516 committed:
Refactor: Rectangle 클래스 프로퍼티 및  하위타입 생성자 수정

- 외부에서 Rectangle 속성에 접근할 수 있도록 private 선언한 부분 제거
- Rectangle.Color의 생성자에서 UIColor과 연동 가능하도록 r,g,b를 모두 255로 나눈 값을 저장

junu0516 committed:
Add: Plane.swift 생성 및 Plane 구조체 구현

- Rectangle 객체배열 프로퍼티, 배열 내 객체 추가 메소드, 서브스크립트 우선 구현

junu0516 committed:
Add: CanvasView.swift 추가 및 UIView 클래스 생성

junu0516 committed:
Add: GeneratorView.swift, StylerView.swift 생성 및 UIView 클래스 생성

junu0516 committed:
Chore: 프로젝트, 스토리보드 관련 파일 수정

junu0516 committed:
Chore: ViewController.swift 위치 변경

- 프로젝트 하위에 Controller 폴더 만들어서 그 밑으로 이동

junu0516 committed:
Chore: Main.storyboard 수정

 - 스토리보드에 UIView 3개 추가하고, 미리 만들어둔 UIView 클래스에 각각 연결

junu0516 committed:
Feat: CanvasView 클래스 UIView 설정 추가

- CanvasView 규격, 배경색 설정
- 서브 뷰로 사각형 추가 버튼 추가
- 버튼 배경색, 타이틀 문자, 테두리 스타일 설정

junu0516 committed:
Refactor: CanvasView 클래스 생성자 수정

- 배경색, 하위 버튼에 넣을 액션 클로저 모두 생성자로 받도록 수정
- CanvasView의 하위 요소에 대한 설정은 모두 해당 클래스 내에서 처리하도록 함

junu0516 committed:
Feat: CanvasView 내부 버튼 스타일 수정

- 버튼 테두리 검정색 둥근 스타일로 변경

junu0516 committed:
Feat: StylerView 클래스 UIView 설정 추가

- UIView 배경, 사이즈 설정
- 서브 뷰로 들어갈 텍스트라벨, 슬라이더 추가

junu0516 committed:
Feat: ViewController에 CanvasView, StyleView 서브 뷰로 추가

junu0516 committed:
Chore: 스토리보드, 프로젝트 파일 등 수정

junu0516 committed:
Feat: CanvasView의 버튼 클릭시 사각형 그리는 기능 추가

- 버튼 클릭시 뷰컨트롤러의 drawRectangle() 메소드 실행
- Rectangle 인스턴스 생성해서 여기에 맞는 UIView 인스턴스 생성
- 생성한 UIView를 Plane에 추가한 후, CanvasView의 서브 뷰로 추가

junu0516 committed:
Refactor: Plane 구조체 CustomStringConvertible 프로토콜 채택

- description 리턴값 정의 후 뷰컨트롤러에서 디버깅

junu0516 committed:
Refactor: Rectangle.Color description 수정, CanvasView 하위 프로퍼티 수정

- r,g,b 값 각각 255다시 곱해서 정수로 변환해서 리턴
- 뷰컨트롤러에서 읽을 수 있도록 generatingButton 접근 제어를 private (set) 으로 수정

junu0516 committed:
Refactor: Rectangle.Alpha 저장 프로퍼티 리턴 타입 변경

- Int에서 10으로 나눈 Float 값을 리턴하도록 변경

junu0516 committed:
Feat: 사각형 선택시 배경색, 투명도 StylerView에 나오도록 기능 추가

- ViewController에 GestureRecognizer 추가
- x,y 좌표값으로 Rectangle 찾을 수 있도록 Plane에 해당 서브스크립트 추가
- 입력받은 rgb값을 16진수로 변환해서 StylerView의 UILabel에 값 업데이트

junu0516 committed:
Feat: 사각형 터치했을 때 테두리 선 표시

- CanvasView에서 선택한 사각형의 x,y값을 가지고 해당되는 서브 뷰 리턴하는 서브스크립트 추가
- CanvasView에서 현재 선택된 서브뷰를 저장할 프로퍼티 변수 추가
- ViewController에서 사각형 선택했을 때 해당 뷰의 테두리 색 변경하는 메소드 호출

junu0516 committed:
Add: ViewMutable.swift, ModelMutable.swift

- 이벤트 발생시 UIView, Model을 각각 변경시킬 프로토콜 추가

junu0516 committed:
Refactor: Plane 구조체 프로퍼티 추가 및 서브스크립트 수정

- 내부에 가지고 있던 인덱스 배열 및 딕셔너리의 키 타입을 Rectangle.Id로 일괄 변경
- Rectangle.Id에 Hashable 프로토콜 채택 추가
- 선택된 사각형 뷰에 해당하는 모델 아이디를 기억하기 위해 selectedRectangleId 프로퍼티 추가

junu0516 committed:
Refactor: ModelMutable, ViewMutable 프로토콜 메소드 수정

- View를 바꾸는건지, Model을 바꾸는건지 좀 더 명확히 할 수 있도록 메소드명 변경
- ModelMutable.changeRectangleModelAlpha에서 x,y 좌표값 필요 없으므로 파라미터에서 제거

junu0516 committed:
Feat: StylerView에서 선택된 뷰의 투명도 변경했을 때 모델의 Rectangle.Alpha값도 변경

- StylerView에서 투명도 바꾸는 액션 일어나면 ViewController이 CanvasView의 투명도를 변경하는 메소드 호출
- CanvasView의 투명도 바꾸는 메소드가 호출되면 ViewController은 선택된 뷰에 해당하는 모델의 Rectangle.Alpha값을 변경하는 메소드 호출

junu0516 committed:
Feat: StylerView에서 배경색 버튼 누르면 선택된 사각형 색 바뀌도록 하기

- StylerView의 버튼 액션이 발생하면 ViewController 메소드 호출
- ViewController은 CanvasView의 해당 사각형의 색상을 변경하는 메소드 호출
- ViewController은 다시 Plane에서 해당 사각형 모델의 정보를 변경하는 메소드 호출

junu0516 committed:
Add: PlaneTest.swift 추가 및 테스트 메소드 작성

junu0516 committed:
Feat: CanvasView 내부의 빈 공간 터치했을 때, 선택 해제 기능 추가

- 빈 공간 터치하면 CanvasView의 사각형 선택효과 해제
- StylerView의 배경색, 투명도 정보 모두 초기화
- Plane 내부의 selectedRectangleId 값을 nil로 변경

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,137 @@
+    private var stylerView: StylerView?
+    private var rectangleFactory: RectangleFactory = RectangleFactory()
+    private var plane: Plane = Plane()
+    var gestureRecognizer = UITapGestureRecognizer()
```
UITapGestureRecognizer가 속성으로 있어야 할까요? 어디에 추가하고 나면 따로 없어도 되지 않을까요? 
특히 이렇게 속성에 바로 초기화를 하는 경우는 init과 달리 초기화가 되기 때문에 메모리만 차지하거나 꼭 사용하지 않는 경우도 있습니다.

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,137 @@
+    private var logger: Logger = Logger()
+    private var canvasView: CanvasView?
+    private var stylerView: StylerView?
+    private var rectangleFactory: RectangleFactory = RectangleFactory()
```
RectangleFactory 같은 팩토리는 보통 인스턴스를 두지 않고 메소드 하나로 호출하고 끝나는 경우가 많습니다

junu0516 replied:
static 메소드로 호출하는 방식을 적용해보겠습니다!

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,137 @@
+                                    buttonActionClosure: self.drawRectangle,
+                                    viewController: self)
+        self.canvasView = canvasView
+        self.view.addSubview(canvasView)
```
CanvasView라는 게 있다면 그걸 self.view로 써보는 것도 고려해보세요

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,137 @@
+                                                alpha: CGFloat(rectangle.alpha.opacity))
+        
+        plane.addRectangle(rectangle)
+        canvasView.insertSubview(rectangleView, belowSubview: canvasView.generatingButton)
```
drawRectangle() 이름으로는 그림을 그리는 것처럼 보이네요
뷰 관련 API 중에서 draw-로 시작하는 메소드에서 뷰를 그리는 작업이 있어서 헷갈릴 수 있습니다
이 메소드가 이벤트가 발생해서 모델로 들어가는 역할인지
모델에서 값이 처리가 된 이후에 뷰로 가는 흐름의 역할인지 구분해보세요
그걸 나누는 게 중요합니다.

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,137 @@
+        let opacity = rectangle.alpha.opacity
+        stylerView.updateRectangleInfo(r: r, g: g, b: b, opacity: opacity)
+        canvasView.selectTappedRectangle(subView: rectangleView)
+        self.plane.selectRectangle(id: rectangle.id)
```
제스처에서 이벤트를 받은 경우에도 이렇게 구분해보세요. 지금은 입력과 출력을 모두 하고 있는 셈입니다. 
plane에게 터치한 좌표를 전달해주고 특정한 rect가 선택되는 흐름만 처리하고
특정한 rect가 선택된 것을 알려주면 그것 받아서 (rect에 대한 view를 매칭해서) canvasView에 선택된 뷰 테두리를 그리는 흐름으로 구분해보세요

이렇게 하나의 메소드에서 다 처리하면,
다른 이벤트로 객체를 선택하거나, 다른 이벤트로 객체가 그려져야 할 때 동일한 코드가 또 생기거나 중복됩니다. 
특히 뷰에 보여지는 방식이 달라지면 출력하는 부분만 바뀌어야 합니다.

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,137 @@
+    func changeRectangleModelColor(rgb: [Double]) {
+        guard let selectedRectangleId = self.plane.selectedRectangleId else { return }
+        guard let rectangle = self.plane[selectedRectangleId] else { return }
+        rectangle.backgroundColor = Rectangle.Color(r: rgb[0]*255, g: rgb[1]*255, b: rgb[2]*255)
```
여기도 흐름을 모델로 가는 흐름과 뷰를 바꾸는 흐름을 구분해보세요
속성 화면에서 들어오는 이벤트라면 모델까지만 들어가도록 하고 끝내보세요

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -0,0 +1,66 @@
+    }
+    
+    mutating func selectRectangle(id: Rectangle.Id){
+        self.selectedRectangleId = id
```
Rectangle.Id를 넘겨주는 게 아니라 좌표를 받아서 plane에서 찾도록 하도록 해보세요

godrm commented on DrawingApp/DrawingApp/Model/Rectangle.swift:
```diff
@@ -2,6 +2,27 @@ import Foundation
+        }
+        
+        init(){
+            var asciiValues: [Int] = []
+            asciiValues.append(contentsOf: Array(97...122))
+            asciiValues.append(contentsOf: Array(48...57))
+            
+            var characters: [Character] = []
+            while(characters.count != 9){
+                guard let value = UnicodeScalar(asciiValues[Int.random(in: 0..<asciiValues.count)]) else { continue }
+                let character = Character(value)
+                characters.append(character)
+            }
+            self.idValue = "\(String(characters[0...2]))-\(String(characters[3...5]))-\(String(characters[6...8]))"
```
만약 Id를 테스트 가능하게 만들려면 지금 init()으로는 랜덤한 요소를 포함하고 있어서 힘들겠네요
init도 테스트 가능한 함수로 만들려면 어떻게 하는게 좋을까요?

godrm commented on DrawingApp/DrawingApp/Model/Rectangle.swift:
```diff
@@ -29,17 +50,27 @@ class Rectangle: CustomStringConvertible{
-            self.r = r
-            self.g = g
-            self.b = b
+            self.r = adjustRange(value: r)/255
+            self.g = adjustRange(value: g)/255
+            self.b = adjustRange(value: b)/255
```
타입을 생성할 때 어떤 식으로 구현할지를 선택해야 하는데,  
• 범위를 벗어난 값이 들어와도 범위로 잘라서 만들어 줄 것인가  
• 혹은 아예 만들어지지 않도록 예외처리를 할 것인가  
• 아예 넣지도 못하도록 할 것인가  
여러가지 방법이 있습니다. 그 중에서도 개발자들이 실수를 줄일 수 있고, 런타임에 실행하지 않고 컴파일 타임에 점검할 수 있으면 가장 좋습니다.  
여기서는 어떻게 처리하고 있는걸까요? 이걸 사용하는 다른 사람이 어떤 부분을 기대하고 실수하고 확인하게 될까요? 그걸 생각해보세요

godrm commented on DrawingApp/DrawingApp/View/CanvasView.swift:
```diff
@@ -0,0 +1,92 @@
+
+class CanvasView: UIView{
+    
+    private var viewController: ModelMutable?
```
뷰가 프로토콜이긴 해도 viewController를 참조하는 것은 좋지 않습니다. 
순환 참조가 되서 메모리에 남는 주요한 요소가 됩니다

junu0516 replied:
혹시 말씀하신 문제가 강한 순환참조로 인한 것이라면, viewController - canvasView가 서로를 참조할 때 한쪽을 약한 참조로 바꾸는 것도 문제가 될까요?

godrm commented on DrawingApp/DrawingApp/View/CanvasView.swift:
```diff
@@ -0,0 +1,92 @@
+        super.init(frame: frame)
+        self.backgroundColor = backGroundColor
+        setGeneratingButton()
+        setGeneratingButtonAction(buttonActionClosure: buttonActionClosure)
```
CanvasView라면 생성하는 버튼과 분리해서 딱 요소들만 그리는 뷰로 만드는 건 어때요?

godrm commented on DrawingApp/DrawingApp/View/CanvasView.swift:
```diff
@@ -0,0 +1,92 @@
+    }
+        
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
```
항상 뷰를 만들 때는 지정 생성자들을 일관되게 만들어주는 게 좋습니다 
CanvasView를 다른 사람이 쓰다가 init(coder:)를 호출하게 되면 전혀 동작하지 않을테니까요

godrm commented on DrawingApp/DrawingApp/View/CanvasView.swift:
```diff
@@ -0,0 +1,92 @@
+    }
+    
+    subscript(x: Double, y: Double)-> UIView?{
+        func isTappedPointInsideTheRange(view: UIView)-> Bool{
+            if((x <= view.frame.maxX && x >= view.frame.minX) && (y <= view.frame.maxY && y >= view.frame.minY)){
+                return true
+            }
+            return false
+        }
+        
+        for view in self.subviews{
+            if(isTappedPointInsideTheRange(view: view)){
+                return view
+            }
+        }
+        
+        return nil
```
이런 역할을 하는걸 hitTest() 라고 합니다
그런데 이 subscript가 왜 필요한지 모르겠습니다. 이미 탭 제스처가 뷰를 알려주지 않던가요?
그리고 뷰를 선택하는 건 좌표로 선택하는 게 아니라, 좌표를 가진 모델을 찾으면 그 모델과 매칭되는 뷰를 선택하도록 해보세요 
뷰는 무언가 찾고 비교하는 로직을 갖기 보다는 이벤트로 입력하거나 지정된 것을 출력하는 역할에 집중해보세요

godrm commented on DrawingApp/DrawingApp/View/GeneratorView.swift:
```diff
@@ -0,0 +1,7 @@
+import Foundation
+import UIKit
+
+
+class GeneratorView: UIView{
+    
+}
```
사용하지 않는 클래스 같네요

godrm commented on DrawingApp/DrawingApp/View/CanvasView.swift:
```diff
@@ -0,0 +1,92 @@
+        }
+    }
+    
+    func selectTappedRectangle(subView: UIView){
+        if let selectedRectangleView = self.selectedRectangleView {
+            selectedRectangleView.layer.borderWidth = 0
+        }
+        subView.layer.borderWidth = 2
+        self.selectedRectangleView = subView
+    }
+    
+    func changeSelectedRectangleOpacity(opacity: Int){
+        guard let viewController = self.viewController else { return }
+        guard let selectedRectangleView = selectedRectangleView else { return }
+        selectedRectangleView.alpha = CGFloat(opacity) / 10
+        viewController.changeRectangleModelAlpha(opacity: opacity)
+    }
+    
+    func changeSelectedRectangleColor(rgb: [Double]){
+        guard let selectedRectangleView = selectedRectangleView else { return }
+        selectedRectangleView.backgroundColor = UIColor(red: rgb[0], green: rgb[1], blue: rgb[2], alpha: 1)
+    }
```
여기 메소드들이 출력용인지, 입력용인지 구분이 잘 안됩니다
selectTappedRectangle()는 선택한 뷰를 표시하는 출력인 것 같은데,
changeSelectedRectangleOpacity()도 출력이라면 이 메소드 자체를 뷰 컨트롤러가 불렀을 것 같은데  `viewController.changeRectangleModelAlpha(opacity: opacity)` 다시 뷰 컨트롤러를 호출해줘서 헷갈립니다.

junu0516 replied:
화면에서 입력이 일어나면, 해당 뷰에서 바로 뷰를 수정처리하고 모델 값을 변경하기 위해 뷰컨트롤러를 호출하는 과정을 의도했습니다. 위에 피드백해주신대로 뷰 자체에서 수정하는 로직을 바로 처리하지 말고 뷰컨트롤러를 매개하도록 수정해야 더 말끔해질 것 같네요

godrm commented on DrawingApp/DrawingApp/View/StylerView.swift:
```diff
@@ -0,0 +1,96 @@
+    }
+    
+    private func setColorChangeAction(){
+        self.rectangleColorValueField.addAction(UIAction(title: ""){_ in
+            let newColor = UIColor(red: CGFloat.random(in: 0...1),
+                                   green: CGFloat.random(in: 0...1),
+                                   blue: CGFloat.random(in: 0...1),
+                                   alpha: 1)
+            guard let rgb = newColor.cgColor.components else { return }
+            guard let viewController = self.viewController else { return }
+            let newHexString = "#\(String(Int(rgb[0]*255), radix: 16))\(String(Int(rgb[1]*255), radix: 16))\(String(Int(rgb[2]*255), radix: 16))"
+            self.rectangleColorValueField.backgroundColor = newColor
+            self.rectangleColorValueField.setTitle(newHexString, for: .normal)
```
StyleView에서도 이런 로직을 처리하지 마시고, 버튼을 눌렸다는 이벤트를 전달하기만 하는 게 좋습니다
그걸 눌렀을 때 랜덤으로 컬러를 뽑을지, 어떤 형태로 표시할 지는 다른 곳에서 처리하는 게 좋습니다

godrm reviewed (changes requested):
입력을 담당하는 뷰에서는 컨트롤러로 이벤트를 보내고, 거기서 로직을 처리하기 위한 모델에 값을 바꾸도록만 해보세요
반대로 모델에서 컨트롤러로 알려주면 출력을 위해서 표시하면 됩니다.
그걸 구분하도록 다음 피드백을 확인해서 개선해보세요

junu0516 committed:
Refactor: ViewController, RectangleFactory 클래스 수정

- ViewController 내부의 gestureRecognizer 속성 제거
- 굳이 뷰컨트롤러 속성으로 두지 않고 setGestureRecognizer() 메소드 내부에서 객체 생성해서 뷰로 전달하는 정도로 끝냄
- ViewController 내부의 rectangleFactory 속성 제거
- RectangleFactory의 메소드를 static으로 변경하여,  뷰컨트롤러에서 팩토리의 인스턴스를 만들지 않고 필요할 때만 static method를 호출하고 끝내는 식으로 변경

junu0516 committed:
Refactor: ViewController 클래스 내부 메소드명 변경

- drawRectangle -> createRectangle로 변경

junu0516 committed:
Refactor: ViewController - StylerView & CanvasView 간 참조 수정

- View가 내부에서 프로토콜 타입의 ViewController을 약한 참조하도록 수정

junu0516 committed:
Refactor: ViewMutable, ModelMutable 이름 변경

- 각각 StylerViewDelegate, CanvasViewDelegate로 좀 더 명확하게 프로토콜명 변경
- CanvasView, StylerView에서 해당 프로토콜 타입 참조할 때도 delegate로 변수명 변경
- 파일명 프로토콜명에 맞게 바꾼 후 View 폴더 하위로 이동

junu0516 committed:
Refactor: ViewController이 하위 UIView 타입 속성 강한 참조하도록 변경

junu0516 committed:
Add: PlaneDeleagte.swift, PlaneDelegate protocol 추가

junu0516 committed:
Refactor: ViewController PlaneDelegate 채택 및  Plane 구조체 delegate 속성 추가

junu0516 committed:
Refactor: ViewController의 createRectangle 메소드를 MVC 구조에 맞춰서 흐름 분기

- CanvasView의 버튼 액션을 delegate의 creatingRectangleRequested()를 호출하여, 뷰컨트롤러로 모델 생성 요청을 보내는 동작으로 수정
- 뷰컨트롤러는 Rectangle 모델 객체 만들어서 plane.addRectangle() 호출
- Plane 객체는 입력받은 Rectangle 객체를 추가한 후 delegate의 addingRectangleComopleted()를 호출하여 뷰컨트롤러에 해당 동작이 끝났음을 알림
- 뷰컨트롤러는 해당 메소드가 호출되면 전달받은 rectangle 객체에 맞는 rectangleView를 생성하여 canvasView에 전달

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -33,9 +38,8 @@ class ViewController: UIViewController{
-                                    viewController: self)
+                                    backGroundColor: .lightGray)
+        canvasView.delegate = self
         self.canvasView = canvasView
         self.view.addSubview(canvasView)
```
이렇게 canvasView를 직접 참조할꺼라면 self.view가 CanvasView가 되는 것도 괜찮은 것 같습니다.

junu0516 replied:
넵 ㅎㅎ 위에 피드백 주신거 보고 수정중이었습니다 다음 단계 진행하면서 수정 마저 진행하겠습니다!

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -46,62 +50,65 @@ class ViewController: UIViewController{
+extension ViewController: UIGestureRecognizerDelegate, PlaneDelegate{
+    
+    func addingRectangleCompleted(rectangle: Rectangle) {
+        if let canvasView = self.canvasView{
+            let rectangleView = UIView(frame: CGRect(x: rectangle.point.x,
```
이런 경우 guard let을 사용하면 early exit이 되고 binding 된 값으로 들여쓰기 없이 표현이 가능합니다.

junu0516 replied:
넵 수정하겠습니다!

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -46,62 +50,65 @@ class ViewController: UIViewController{
+                                                    green: rectangle.backgroundColor.g,
+                                                    blue: rectangle.backgroundColor.b,
+                                                    alpha: CGFloat(rectangle.alpha.opacity))
+            canvasView.insertSubview(rectangleView, belowSubview: canvasView.generatingButton)
```
뷰를 생성하는 것도 Factory가 있으면 어떨까요?

junu0516 replied:
넵 모델처럼 뷰도 서브뷰 만드는 팩토리의 메소드를 호출하는 식으로 수정하겠습니다

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -11,10 +11,13 @@ struct Plane:CustomStringConvertible{
+    weak var delegate: PlaneDelegate?
     
     mutating func addRectangle(_ rectangle: Rectangle){
+        guard let delegate = self.delegate else { return }
         self.rectangles[rectangle.id] = rectangle
         self.rectangleIndex.append(rectangle.id)
+        delegate.addingRectangleCompleted(rectangle: rectangle)
```
델리게이트 객체가 없다고 해서 아예 동작을 처리안 하면 안되고, 맨 아래 호출만 안해야하지 않을까요?

junu0516 replied:
delegate 부분만 if-let으로 호출하는 식으로 변경하겠습니다!

godrm commented on DrawingApp/DrawingApp/View/CanvasView.swift:
```diff
@@ -21,10 +20,9 @@ class CanvasView: UIView{
         if let selectedRectangleView = self.selectedRectangleView,
-           let viewController = self.viewController {
+           let viewController = self.delegate {
             selectedRectangleView.layer.borderWidth = 0
             self.selectedRectangleView = nil
-            
             viewController.clearModelSelection()
```
델리게이트라고 이름을 붙이는 것은 그게 viewController가 아닐 수도 있기 때문입니다. 
바인딩해서 변수명을 정할 때도 그게 꼭 ViewController가 아닐 수 있다고 가정하는 게 좋습니다
델리게이트는 누가 담당하던지 상관없는 역할의 이름일 뿐입니다. 영화에서 배역의 역할, 학급 반장이나 주번 같은 역할이라고 생각하세요 
그 역할이 없다고 해서 동작 자체를 안하면 안되고, 그 객체의 메소드를 호출하는 것만 제외해야 합니다.

junu0516 replied:
아하 이름 자체도 ViewController임을 직접적으로 드러내지 않아야 되는군요! 변수명도 같이 수정하겠습니다

godrm reviewed (approved):
weak 참조 정도로 delegate를 가지는 정도는 괜찮습니다 (순환 참조도 안 생기니까요)
그렇지만 그게 어떤 객체인지 모르고 있는 게 핵심입니다
아래 피드백처럼 delegate가 없다고 동작을 아예 안하면 안되고, 그게 viewcontroller가 아닐 수도 있다고 가정하는 게 좋습니다. 이제 다음 단계 옵저버 형태로 개선해보시면 될 것 같네요

junu0516 commented:
쉬는 날인데 상세한 피드백 감사합니다! 😄 MVC 흐름에 맞게 메소드 단방향으로 분기하는 부분 수정중이었는데, 다음 단계 진행하면서 수정도 마저 진행하겠습니다!
